proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,820 @@
+  Supervisor.
+- Provide a mechanism which governs a client's access to the token exchange APIs. Not all webapps should have permission
+  to act on behalf of the user with the Kubernetes API of the clusters, so an admin should be able to configure which
+  clients have this permission.
+- Provide a mechanism for requesting access to different aspects of a user identity, especially getting group
+  memberships or not, to allow the admin to exclude this potentially information for clients which do not need it.
+- Support a web UI based LDAP/ActiveDirectory login screen. This is needed to avoid having webapps handle the user's
+  password, which should only be seen by the Supervisor and the LDAP server. However, the details of this item have been
+  split out to a separate proposal document.
+- Client secrets should be stored encrypted or hashed, not in plain text.
+
+Non-goals for this proposal:
+
+- Pinniped's scope is to provide authentication for cluster users. Providing authentication for arbitrary users to
+  arbitrary webapps is out of scope. The only proposed use case is providing the exact same identities that are provided
+  by using Pinniped's `kubectl` integration, which are the developers/devops/admin users of the cluster.
+- Supporting any OAuth/OIDC flow other
+  than [OIDC authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth).
+- Implementing any self-service client registration API. Clients will be registered by the Pinniped admin user.
+- Implementing a consent screen. This would be clearly valuable but will be left as a potential future enhancement in
+  the interest of keeping the first draft of this feature smaller.
+- Management (ie creation & rotation) of client credentials on the operator's behalf. This will be the operator's
+  responsibility.
+- Orchestration of token exchanges on behalf of the client. Webapps which want to make calls to the Kubernetes API of
+  clusters acting as the authenticated user will need to perform the rest of the token and credential exchange flow that
+  it currently implemented by the Pinniped CLI. Providing some kind of component or library to assist webapp developers
+  with these steps might be valuable but will be left as a potential future enhancement.
+
+### Specification / How it Solves the Use Cases
+
+This document proposes supporting a new Custom Resource Definition (CRD) for the Pinniped Supervisor which allows the
+admin to create, update, and delete OIDC clients for the Supervisor.
+
+#### API Changes
+
+##### Configuring clients
+
+An example of the new CRD to define a client:
+
+```yaml
+apiVersion: oauth.supervisor.pinniped.dev/v1alpha1
+kind: OIDCClient
+metadata:
+  name: my-webapp-client
+  namespace: pinniped-supervisor
+spec:
+  id: my-webapp
+  secretNames:
+    - my-webapp-client-secret-1
+    - my-webapp-client-secret-2
+  allowedRedirectURIs:
+    - https://my-webapp.example.com/callback
```

Comment:
Perhaps we should limit this to a single string instead of a list to discourage the use of the same client across multiple web apps.

Reply:
That would be different to most other client definitions, wouldn't it? (not that it's a reason not to do so)

Reply:
> @absoludity I'm not sure why, but GitHub won't let me reply to your comment which said "That would be different to most other client definitions, wouldn't it? (not that it's a reason not to do so)".
> 
> I'm not sure what you mean by that comment. Could you explain? Thanks!

I meant simply that other providers that I've used have always allowed multiple redirect URIs (though might just be the ones I've used). It'd be a surprise to find I couldn't do so, but may not be a bad thing :) (My main reason for using multiple redirect URLs is usually because I've got a client setup for use in different development projects, so not a good reason to allow it here of course).

Reply:
Ah, yes, thanks for explaining. We're planning to allow multiple redirect URIs to be configured.